### PR TITLE
Implemented R437 and R502 of Fortran2008 standard (#247)

### LINKED ISSUE #247: [fparser2][Fortran2008] Missing attribute CONTIGUOUS
Fortran2008 introduced the `CONTIGUOUS` attribute in [R437](https://j3-fortran.org/doc/year/10/10-007r1.pdf#page=84), [R502](https://j3-fortran.org/doc/year/10/10-007r1.pdf#page=105) which fparser2 cannot handle currently:
```
>>> from fparser.two.parser import ParserFactory
>>> from fparser.common.readfortran import FortranStringReader
>>> code = '''
... module contiguous_ptr
... type some_type
...   integer, pointer, contiguous :: ptr(:) => null()
... end type some_type
... end module contiguous_ptr
... '''
>>> parser = ParserFactory().create(std='f2008')
>>> reader = FortranStringReader(code)
>>> ast = parser(reader)
Traceback (most recent call last):
  File "...src/fparser/two/Fortran2003.py", line 236, in __new__
    return Base.__new__(cls, string)
  File "...src/fparser/two/utils.py", line 407, in __new__
    raise NoMatchError(errmsg)
fparser.two.utils.NoMatchError: at line 4
>>>  integer, pointer, contiguous :: ptr(:) => null()
```

I could attempt to provide a PR patching this feature but I need some hand holding in how to do this in the desired way as it affects multiple things:
- `Fortran2003.Component_Attr_Spec` needs the additional match against `CONTIGUOUS`? 
- `Fortran2003.Attr_Spec`, too. But here the pattern is defined in `pattern_tools`.
- Whether/how to enforce [C448](https://j3-fortran.org/doc/year/10/10-007r1.pdf#page=84) and [C530](https://j3-fortran.org/doc/year/10/10-007r1.pdf#page=111)?

Additionally, R437 and R503 introduce the attribute `CODIMENSION`. Would it be necessary to add this as well, when I'm "touching the relevant rule"?

### PR DESCRIPTION
**Pull request by reuterbal**
This is an attempt to add the missing `CONTIGUOUS` and `CODIMENSION` attributes introduced in Fortran2008. This is to fix #247

### CODE REVIEW

File: src/fparser/two/Fortran2008.py [hunk 1]
```diff
@@ -96,6 +101,241 @@ class Program_Unit(Program_Unit_2003):  # R202
     subclass_names.append("Submodule")
 
 
+class Component_Attr_Spec(STRINGBase):  # R437
+    """
+:F08R:`437`
```
Comment (rupertford) [inline src/fparser/two/Fortran2008.py]: Could you please use the newer Formatting for docstrings (take a look at the other routines in the Fortran2008.py classes for examples).

File: src/fparser/two/Fortran2008.py [hunk 2]
```diff
@@ -96,6 +101,241 @@ class Program_Unit(Program_Unit_2003):  # R202
...
+    use_names = []
+
+    @staticmethod
+    def match(string):
+        '''Implements the matching for component attribute specifications.
+
+        :param str string: the string to match as an attribute.
+        :return: None if there is no match, otherwise an object with the \
+            matched attribute.
+        :rtype: None, Component_Attr_Spec, Codimension_Attr_Spec, \
```
Comment (rupertford) [inline src/fparser/two/Fortran2008.py]: classes need to be specified as `:py:class:...`. Also, I think it is either None or a 1-tuple, containing an instance of one of the matched classes that is returned.

File: src/fparser/two/Fortran2008.py [hunk 3]
```diff
@@ -96,6 +101,241 @@ class Program_Unit(Program_Unit_2003):  # R202
...
+            Access_Spec, Dimension_Component_Attr_Spec
+        '''
+        return STRINGBase.match(
+            ['ALLOCATABLE', 'CONTIGUOUS', 'POINTER'], string)
+
+
+class Codimension_Attr_Spec(WORDClsBase):  # R502.d
+    """
+    <codimension-attr-spec> = CODIMENSION <lbracket> <coarray-spec> <rbracket>
```
Comment (rupertford) [inline src/fparser/two/Fortran2008.py]: Again, could you please modify the docstring format. I think this applies to subsequent changes as well but I stop mentioning it now :-)

File: src/fparser/two/Fortran2008.py [hunk 4]
```diff
@@ -96,6 +101,241 @@ class Program_Unit(Program_Unit_2003):  # R202
...
+    subclass_names = []
+    use_names = ['Coarray_Bracket_Spec']
+
+    @staticmethod
+    def match(string):
+        '''Implements the matching for the CODIMENSION attribute.
+
+        :param str string: the string to match as the attribute.
+        :return: None if there is no match, otherwise the matched object.
+        :rtype: None, Codimension_Attr_Spec
```
Comment (rupertford) [inline src/fparser/two/Fortran2008.py]: Please see previous comment about return types. I think this applies to subsequent classes as well but will stop mentioning it now.

File: src/fparser/two/Fortran2008.py [hunk 5]
```diff
@@ -96,6 +101,241 @@ class Program_Unit(Program_Unit_2003):  # R202
...
+    @staticmethod
+    def match(string):
+        '''Implements the matching for deferred coarray shape specification.
+
+        :param str string: the string to match as deferred shape.
+        :return None if there is no match, otherwise the matched object.
+        :rtype: None, Deferred_Coshape_Spec
+        '''
+        if string == ':':
+            return None, None
```
Comment (rupertford) [inline src/fparser/two/Fortran2008.py]: I think pylint like tuples to be specified within brackets these days i.e. `(None, None)`

File: src/fparser/two/Fortran2008.py [hunk 6]
```diff
@@ -96,6 +101,241 @@ class Program_Unit(Program_Unit_2003):  # R202
...
+        return
```
Comment (rupertford) [inline src/fparser/two/Fortran2008.py]: I think pylint does not like a mix of return value and no return value. So please just return `None` here.

File: src/fparser/two/Fortran2008.py [hunk 7]
```diff
@@ -96,6 +101,241 @@ class Program_Unit(Program_Unit_2003):  # R202
...
+        :rtype: None, Explicit_Coshape_Spec, Coshape_Spec_List
+        '''
+        if not string.endswith('*'):
+            return
+        line = string[:-1].rstrip()
+        if not line:
+            return None, None
+        if line.endswith(':'):
+            line, repmap = string_replace_map(line[:-1].rstrip())
+            i = line.rfind(',')
```
Comment (rupertford) [inline src/fparser/two/Fortran2008.py]: Please check all changes with pycodestyle and pylint as I think `i` will cause them to complain.

File: src/fparser/two/Fortran2008.py [hunk 8]
```diff
@@ -96,6 +101,241 @@ class Program_Unit(Program_Unit_2003):  # R202
...
+                Lower_Cobound(repmap(line[i+1:].lstrip()))
+        if not line.endswith(','):
+            return
+        line = line[:-1].rstrip()
+        return Coshape_Spec_List(line), None
+
+    def tostr(self):
+        '''Converts the explicit coarray shape specification to string.'''
+        s = ''
+        if self.items[0] is not None:
```
Comment (rupertford) [inline src/fparser/two/Fortran2008.py]: I think `if self.items[0]:` would work here

File: src/fparser/two/Fortran2008.py [hunk 9]
```diff
@@ -96,6 +101,241 @@ class Program_Unit(Program_Unit_2003):  # R202
...
+        including the square brackets.
+
+        :param str string: the string to match as the specification.
+        :return: None if there is no match, otherwise the matched object.
+        :rtype: None, Coarray_Bracket_Spec
+        '''
+        return BracketBase.match('[]', Coarray_Spec, string)
+
+
+class Attr_Spec(STRINGBase):  # R502
```
Comment (rupertford) [inline src/fparser/two/Fortran2008.py]: I think it would be simpler to implement this as a subclass of Fortran2003.py's R503 (see Program_Unit in Fortran2008.py as an example).

**Comment by rupertford** [comment]
Back to @reuterbal

File: src/fparser/two/Fortran2008.py [hunk 10]
```diff
@@ -96,6 +101,241 @@ class Program_Unit(Program_Unit_2003):  # R202
     subclass_names.append("Submodule")
 
 
+class Component_Attr_Spec(STRINGBase):  # R437
```
[the remark this answers is not in the document]
Comment (reuterbal) [reply src/fparser/two/Fortran2008.py]: Wouldn't that still require re-implementing the match-method in the F08 class? Since `match` is a staticmethod it would require to refer explicitly to the class variable (e.g., `Component_Attr_Spec.attributes`) and thus overwriting it in a subclass would not change the behaviour in the match method.

I was able to get it to work by changing `match` to a classmethod in the F03 class and use `cls.attributes` in the matching, i.e.,
```python
class Component_Attr_Spec(STRINGBase):
    ...
    attributes = [...]
    @classmethod
    def match(cls, string):
        return STRINGBase.match(cls.attributes, string)
```
Then I can setup the F08 class as
```python
class Component_Attr_Spec(Component_Attr_Spec_2003):
    subclass_names = Component_Attr_Spec_2003[:]
    subclass_names.append('Codimension_Attr_Spec')
    Component_Attr_Spec_2003.attributes.append('CONTIGUOUS')
    match = Component_Attr_Spec_2003.match
```
Interestingly, it doesn't work if I extend `attributes` in the same way as `subclass_names`. This does somewhat exceed my knowledge about edge cases of Python inheritance...

File: src/fparser/two/Fortran2008.py (hunk 9 above)
Comment (reuterbal) [reply src/fparser/two/Fortran2008.py]: Makes sense for consistency but would only change the definition of `subclass_names`. The match is done using a pattern and requires reimplementation for F08.

File: src/fparser/two/Fortran2008.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (reuterbal) [reply src/fparser/two/Fortran2008.py]: OK, there's good and bad news:

After some discussion and digging with @mlange05 we identified the cause for that behaviour. In [utils.py:323](https://github.com/stfc/fparser/blob/master/src/fparser/two/utils.py#L323) the `match` method is obtained by looking into the `cls.__dict__`:
``` match = cls.__dict__.get('match')```
instead of using `getattr`
```match = getattr(cls, 'match') if hasattr(cls, 'match') else None```
which breaks native inheritance. Thus, without explicitly implementing the match method in the subclass (which above assignment does), there is no 'match' in `__dict__`. And with the assignment, `cls` is not the subclass but the parent. (Interesting Python feature BTW).

With that said, I can extend `attributes` in the same way as `subclass_names` and leave out the definition of `match` if the F03 match is turned into a classmethod and utils.py is altered in the way above.

The bad news: this change breaks other tests (src/fparser/two/tests/fortran2003/test_defined_operators_r703_r723.py).

The good news: my knowledge about polymorphism in Python wasn't so far off after all... ;-)

Is there a specific reason that `match` is obtained in the above way?

[the remark this answers is not in the document]
Comment (reuterbal) [reply src/fparser/two/Fortran2008.py]: Following up on this: The failing tests appear when trying to match `Defined_Binary_Op` and `Defined_Unary_Op` instead of one of the subclasses implementing these operator classes. If I understand correctly, the idea is that these should not actually be able to match successfully but only via their subclasses. 
Above changes in utils.py yield a match method for both classes (from STRINGBase) with wrong number of parameters (expects `pattern` as additional argument). Tests can be repaired when setting `match=None` in both classes, which could be a viable way of indicating "This class should not match".

**Comment by reuterbal** [comment]
@rupertford I don't know if you received emails about the above comments: https://github.com/stfc/fparser/pull/249#discussion_r388215617
If you find time the time, could you share some thoughts?

File: src/fparser/two/Fortran2008.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/Fortran2008.py]: Thanks for all this investigation. When I suggested the change it was meant to make things simpler, not lead you into a warren of problems :-). You are, of course, right about not being able to do what I suggested for a static method - sorry about that.

In answer to your question about getting the match method using __dict__, this part of the code is inherited and not previously examined by me, so I don't know why it was implemented this way. I can't see any reason why it should stay as it is and I prefer the ability for subclasses to override match methods so would be very happy for you to make the change to use `hasattr` and `getattr` if you don't mind.

As for the failing test, I don't think the two offending classes should inherit from STRINGBase as they do not use this functionality. If they are changed to inherit from Base instead then the tests are OK (as it has no match function). If you are also happy to make that change as well, that would be great.

Also, many thanks for finding the very nice `classmethod` solution, it could be very useful as we inherit our way towards Fortran2018!

**Comment by rupertford** [comment]
> @rupertford I don't know if you received emails about the above comments: [#249 (comment)](https://github.com/stfc/fparser/pull/249#discussion_r388215617)
> If you find time the time, could you share some thoughts?

Yes @reuterbal, I had read them and thought it all sounded complicated so went off and did something else, then forgot I was meant to reply :-). I think I now understand what you've done and what the issues are and have made a suggestion as a response to the original comment. Many thanks for the great work.

File: src/fparser/two/Fortran2008.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (reuterbal) [reply src/fparser/two/Fortran2008.py]: Many thanks for your reply! I hope you don't mind me nudging.

Your suggestion sounds great, I'll happily do the changes. Would you like to have them as a separate PR or is it fine in this one?

**Comment by reuterbal** [comment]
I think all requested changes are addressed now.
The change to the lookup of `match` with the accompanying switch from `STRINGBase` to `Base` as parent class is implemented in 6ccab21. If you would like to see this in a separate PR I can easily cherry-pick that out. 
Let me know please what you want to have in regard to testing constraints.

File: src/fparser/two/tests/fortran2008/test_attr_spec.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    obj = tcls('codimension [*]')
+    assert isinstance(obj, Codimension_Attr_Spec), repr(obj)
+    assert str(obj) == 'CODIMENSION [*]'
+
+
+def test_codimension_attr_spec(f2008_create):
+    '''Tests the parsing of codimension attributes.'''
+    tcls = Codimension_Attr_Spec
+    attrs = {
+        'codimension[*]': 'CODIMENSION [*]',
```
Comment (rupertford) [inline src/fparser/two/tests/fortran2008/test_attr_spec.py]: Test mixed case and spaces before/after

Comment (rupertford) [reply src/fparser/two/tests/fortran2008/test_attr_spec.py]: Also test failure to match due to typo or space in wrong location or incorrect brackets e.g. too many, not enough etc.

File: src/fparser/two/tests/fortran2008/test_attr_spec.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+def test_invalid_coshape_spec(f2008_create):
+    '''Tests invalid codimension attributes.'''
+    tcls = Codimension_Attr_Spec
+    attrs = ['codimension[1:5]', 'codimension[3:]', 'codimension[:5]',
+             'codimension[,,]', 'codimension[1, 5, 3 *]']
+    for attr in attrs:
+        with pytest.raises(Fortran2003.NoMatchError):
+            _ = tcls(attr)
+
+    with pytest.raises(Fortran2003.NoMatchError):
```
Comment (rupertford) [inline src/fparser/two/tests/fortran2008/test_attr_spec.py]: deferred_coshape_spec should have its own tests with checks for additional spaces, incorrect text, valid text and two colons.

File: src/fparser/two/tests/fortran2008/test_attr_spec.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        _ = Deferred_Coshape_Spec('')
+
+    attrs = [':', 'a:', ':b']
```
Comment (rupertford) [inline src/fparser/two/tests/fortran2008/test_attr_spec.py]: coshape_spec should have its own test(s) and include valid examples as errors, with and without spaces.

File: src/fparser/two/tests/fortran2008/test_attr_spec.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    for attr, ref in attrs.items():
+        obj = tcls(attr)
+        assert isinstance(obj, tcls), repr(obj)
+        assert str(obj) == ref
+
+
+def test_invalid_coshape_spec(f2008_create):
+    '''Tests invalid codimension attributes.'''
+    tcls = Codimension_Attr_Spec
+    attrs = ['codimension[1:5]', 'codimension[3:]', 'codimension[:5]',
```
Comment (rupertford) [inline src/fparser/two/tests/fortran2008/test_attr_spec.py]: Add tests for case insensitivity and spaces and errors due to typos.

File: src/fparser/two/tests/fortran2008/test_attr_spec.py [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+        'protected': 'PROTECTED', 'save': 'SAVE', 'target': 'TARGET',
+        'value': 'VALUE', 'volatile': 'VOLATILE'}
+    for attr, ref in attrs.items():
+        obj = tcls(attr)
+        assert isinstance(obj, tcls), repr(obj)
+        assert str(obj) == ref
+
+    obj = tcls('dimension(:)')
+    assert isinstance(obj, Fortran2003.Dimension_Attr_Spec), repr(obj)
+    assert str(obj) == 'DIMENSION(:)'
```
Comment (rupertford) [inline src/fparser/two/tests/fortran2008/test_attr_spec.py]: spaces and case tests.

File: src/fparser/two/tests/fortran2008/test_attr_spec.py [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+
+    obj = tcls('codimension [*]')
+    assert isinstance(obj, Codimension_Attr_Spec), repr(obj)
+    assert str(obj) == 'CODIMENSION [*]'
```
Comment (rupertford) [inline src/fparser/two/tests/fortran2008/test_attr_spec.py]: spaces and case tests.

File: src/fparser/two/tests/fortran2008/test_component_attr_spec.py [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+from fparser.two.Fortran2003 import Access_Spec, Dimension_Component_Attr_Spec
+from fparser.two.Fortran2008 import Component_Attr_Spec, Codimension_Attr_Spec
+
+
+def test_pointer_attr(f2008_create):
+    '''Tests the parsing of POINTER attribute.'''
+    tcls = Component_Attr_Spec
+    obj = tcls('pointer')
+    assert isinstance(obj, tcls), repr(obj)
+    assert str(obj) == 'POINTER'
```
Comment (rupertford) [inline src/fparser/two/tests/fortran2008/test_component_attr_spec.py]: This and following tests should also test for extra spaces, case insensitivity and invalid content.

File: src/fparser/two/tests/fortran2008/test_component_attr_spec.py [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+    assert isinstance(obj, tcls), repr(obj)
+    assert str(obj) == 'CONTIGUOUS'
+
+
+def test_dimension_attr(f2008_create):
+    '''Tests the parsing of DIMENSION attribute.'''
+    tcls = Component_Attr_Spec
+    obj = tcls('dimension(a)')
+    assert isinstance(obj, Dimension_Component_Attr_Spec), repr(obj)
+    assert str(obj) == 'DIMENSION(a)'
```
Comment (rupertford) [inline src/fparser/two/tests/fortran2008/test_component_attr_spec.py]: This class includes the keyword DIMENSION and the brackets so test for validity of each of these. 1) spaces, 2) case insensitivity and 3) valid/invalid brackets. Similary for the remaining tests in this file.

File: src/fparser/two/Fortran2008.py [hunk 19]
```diff
@@ -96,6 +103,279 @@ class Program_Unit(Program_Unit_2003):  # R202
     subclass_names.append("Submodule")
 
 
+class Component_Attr_Spec(Component_Attr_Spec_2003):  # R437
+    '''
+    Fortran 2008 rule R437
+    component-attr-spec is access-spec
+                           or ALLOCATABLE
+                           or CODIMENSION <lbracket> coarray-spec <rbracket>
```
[the remark this answers is not in the document]
Comment (reuterbal) [reply src/fparser/two/Fortran2008.py]: Not particularly... `lbracket ... rbracket` is actually `[ ... ]` but I didn't want to use the brackets directly to avoid misinterpretation as optional parts

File: src/fparser/two/Fortran2008.py [hunk 20]
```diff
@@ -96,6 +103,279 @@ class Program_Unit(Program_Unit_2003):  # R202
...
+                 or PARAMETER
+                 or POINTER
+                 or PROTECTED
+                 or SAVE
+                 or TARGET
+                 or VALUE
+                 or VOLATILE
+
+    '''
+    subclass_names = ['Access_Spec', 'Language_Binding_Spec',
```
[the remark this answers is not in the document]
Comment (reuterbal) [reply src/fparser/two/Fortran2008.py]: Absolutely, sorry for missing that.

File: src/fparser/two/pattern_tools.py [hunk 21]
```diff
@@ -472,6 +472,13 @@ def __call__(self, string):
     flags=re.I)
 abs_attr_spec = abs(attr_spec)
 
+attr_spec_f08 = Pattern(
```
[the remark this answers is not in the document]
Comment (reuterbal) [reply src/fparser/two/pattern_tools.py]: Please see the updated version if this is what you had in mind. It's a bit convoluted to read but I tried to make it clear with a comment and indentation: Strip parentheses, split at `|`, add attribute, and then re-assemble the string.

**Comment by reuterbal** [comment]
Thank you. I have implemented the requested changes but not yet looked at testing constraints. However, I have noticed that attrib_spec is not matched correctly when parsing a program, module or subroutine (e.g., example from #247) if one of the newly added keywords is included together with another attribute, e.g.
```
integer, pointer, contiguous :: a
```
fails to match, while
```
integer, contiguous :: a
```
works fine (ignoring that it is nonsense for now).
I guess I'm doing something wrong with generating the `Attrib_Spec_List` but haven't found out what, yet. I'll return to that on Monday.

**Comment by reuterbal** [comment]
Inspecting the initialization of the class list in `parser.py`, it turns out there is no F2008 variant of `Attr_Spec_List` generated (and thus the F2003 variant is used). The reason for that is that there is no class in `Fortran2008.py` which has `Attr_Spec` in their `use_names` or `subclass_names`.

This could be fixed, for example, by creating a corresponding empty subclass of `Type_Declaration_Stmt`:
```
class Type_Declaration_Stmt(Type_Declaration_Stmt_2003):
    pass
```
which then triggers generating a `Attr_Spec_List`.

However, with that I can't get any Declarations to parse successfully anymore and I'm at loss why that is. Do you have any better ideas for that?

**Comment by reuterbal** [comment]
It has been some eventful weeks but not so much on this PR - at least until a couple of hours ago. I bothered @mlange05 for a lengthy remote debugging session and we figured out the underlying issue. Essentially, there are two things relevant which I tried to document cleanly in the docstrings, too:
1. `Type_Declaration_Stmt` and `Data_Component_Def_Stmt` need to be added to `Fortran2008.py` without actually changing the implementation of the F2003-versions. The reason is that only with these classes present, `Attr_Spec_List` and `Component_Attr_Spec_List` are generated due to them being mentioned in the `use_names` properties.
2. Unfortunately, for these two classes the `match` method has to be reimplemented without actually changing anything. The reason is that the call to the parent base class requires `attr_spec_list` to be provided, which would always use the Fortran2003 version otherwise. Sadly, above solution of extending/updating a dedicated property that would then be used in the match method does not work as we have a circular dependency: The relevant property of, e.g., `Type_Declaration_Stmt`, would require, in this case, `Attr_Spec_List` to be already available such that the class is fully defined. However, generating `Attr_Spec_List` is only triggered from the `use_names` property of `Type_Declaration_Stmt`.
If you have any better solution for this problem, I'm happy to change the code.

I will look at testing constraints next.

**Comment by reuterbal** [comment]
Following your suggestion I listed the constraints that mention the rules in parentheses in the docstrings of the F2003 and F2008 classes and created #258 and #259 to implement those. These are some long lists...
Almost all constraints actually "belong" to the parent classes `Type_Declaration_Stmt` and `Data_Component_Def_Stmt` which are only duplicated to trigger generating the lists as mentioned in the previous component.
I tried to mimic the formatting used by Andy in one of the recent commits.

I think it's safe to have a look again @rupertford .

File: src/fparser/two/Fortran2008.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/Fortran2008.py]: The answer was this PR please, but you've already done that :-).

File: src/fparser/two/pattern_tools.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/pattern_tools.py:477]: No, that is great. Thanks. As you've probably noticed I like the idea of code for the new standard explicitly extending the old where possible.

File: src/fparser/two/Fortran2008.py (hunk 9 above)
Comment (rupertford) [reply src/fparser/two/Fortran2008.py]: OK.

File: src/fparser/two/Fortran2003.py [hunk 22]
```diff
@@ -1697,10 +1729,25 @@ class Component_Attr_Spec(STRINGBase):  # R441
...
+        Note that this is implemented as a `classmethod` (not a
+        `staticmethod`), using attribute keywords from the list provided
+        as a class property. This allows expanding this list for
+        Fortran 2008 without having to reimplement the matching.
+
+        :param str string: the string to match as an attribute.
+
+        :return: None if there is no match, otherwise a 1-tuple \
+            containing the matched attribute string.
+        :rtype: `None` or (`str`,)
```
[the remark this answers is not in the document]
Comment (reuterbal) [reply src/fparser/two/Fortran2003.py:1747]: Sorry, that's an oversight. Changed that and also added full path for rtype classes in two instances.

**Comment by reuterbal** [comment]
Thanks @rupertford for looking over the code again. 

I don't mind splitting up the existing tests and renaming test files - was rather quick to do and saves the effort of creating a new issue that someone has to look at again. In that instance I also noticed that there were no explicit tests for `*_Cobound` and some `*_Coshape_Spec` cases. I took the liberty of adding those ;-)

This means, there are now tests for R437, R502, R509-513 within their respective files. However, there are no new F2008 tests for R436 and R501 as those are simply "copied". The same applies to some extent to the subclasses of R437 and R502 that are simply inherited from F2003. How do you want to handle these situations in general? Copy all tests from F2003 or only test changes and add some sanity tests that ensure that existing functionality is still there (without being complete)? (Not that I want to add any of those here, just out of curiosity if you have a strict plan for that).

This applies for example to R437: It includes _access-spec_, _coarray-spec_, and _component-array-spec_ (described in R507, R509, and R439, respectively). _coarray-spec_ is new and tested explicitly, the rest is "only" inherited from F2003. The new tests include a test that makes sure that matching R437 with `DIMENSION` still works and produces R439 (i.e., that this does still work) but without explicitly using R439's class to do the match. Thus this makes sure that in this instance the `subclass_names` were used correctly. If new tests would now be required for everything and not just changes, then strictly speaking there would now also be a need for a `_r439.py` file in Fortran 2008.

Hence my question: To what extent are tests required? I do support the idea of requiring tests whenever code for a rule exists in F2008 (such as R436) but should it test the full spectrum (as in all aspects of all inherited subclasses) or just make sure that it behaves as before by means of some sanity tests? The latter would be relatively easy to add using just a few common cases and then relying on proper checking of edge cases in 2003 tests.

The remaining question for this PR is now, if a comment should be included somewhere pointing to an issue for testing of R436 and R501? Or just open the issue? Or nothing at all? ;-)

**Comment by reuterbal** [comment]
@rupertford Can I just ask for some quick input regarding the extent of required tests before I'll mark this ready for review?

**Comment by rupertford** [comment]
> @rupertford Can I just ask for some quick input regarding the extent of required tests before I'll mark this ready for review?

Sorry. I'm afraid I've neglected fparser recently. I'll take a look ...

**Comment by rupertford** [comment]
OK, so here is what I think, but feel free to disagree and perhaps @arporter might want to comment.

As agreed there should be separate test files per class named in the suggested way. The tests on a class should check the functionality of that class and not be concerned with functionality of ancestors or descendants. So we really are trying to implement unit tests. However, it may be easier to test the functionality of a class by calling it from an ancestor or testing results from a descendant and that is fine.

In the case of replicated code I think we should still test it separately, which presumably would mean replicating tests, unless there is a way of importing tests into another file. Whilst testing one should imply the other is correct it seems safer to me to test both. It also allows us to check that all code is covered.

In the case of slightly modified code I don't think it is enough to just test the modification works. One reason for this is that the match might result in a later match no longer matching so we need to check that this does not happen i.e. later matches are at least still reachable. So again I think the class should be fully tested (either by replication or importing). I would be happy for an issue to be added so we can get this PR onto master.

**Comment by arporter** [comment]
My feeling, without looking at the code itself, is that new code that is in Fortran2008 should be covered by new tests. The existing Fortran2003 tests *should* (but may not) be sufficient to ensure no new bugs are introduced, if that is what the issue is. If they are not sufficient then they should be extended where appropriate. Does that help? Hopefully I've answered the right question...

**Comment by reuterbal** [comment]
Yes, I think that clarifies it. Many thanks for replying so quickly! I'll finish this asap.

**Comment by rupertford** [comment]
23rd April to 2nd June is not quick :-)

**Comment by reuterbal** [comment]
Yeah, I meant after my gentle nudge ;-)

**Comment by reuterbal** [comment]
As written above:
- Split tests into per-rule test files
- Copied tests for R501 from Fortran2003
- Opened #265 about tests for R436 and created an empty test file with a comment about the missing tests and the issue number.

I hope that covers the requested changes. And I think this is good to have another look.
